http/src/main/java/io/micronaut/http/filter/FilterProperties.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2017-2019 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.filter;
+
+import io.micronaut.core.annotation.AnnotationValue;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.util.ArrayUtils;
+import io.micronaut.http.HttpMethod;
+import io.micronaut.http.annotation.Filter;
+
+import java.lang.annotation.Annotation;
+
+/**
+ * Encapsulates all the possible configurations that might be defined in {@link Filter} annotation.
+ */
+@Internal
+public final class FilterProperties {
+
+    public static final FilterProperties EMPTY_FILTER_PROPERTIES = new FilterProperties(
+            new String[0],
+            new HttpMethod[0],
+            new String[0],
+            new Class[0]
+    );
+
+    private final String[] patterns;
+    private final HttpMethod[] methods;
+    private final String[] serviceId;
+    private final Class<? extends Annotation>[] annotationMarkers;
+
+    /**
+     * @param patterns          the patterns this filter should match
+     * @param methods           the methods to match
+     * @param serviceId         the serviceId to match
+     * @param annotationMarkers the annotation markers to match
+     */
+    public FilterProperties(String[] patterns,
+                            HttpMethod[] methods,
+                            String[] serviceId,
+                            Class<? extends Annotation>[] annotationMarkers) {
+        this.patterns = patterns;
+        this.methods = methods;
+        this.serviceId = serviceId;
+        this.annotationMarkers = annotationMarkers;
+    }
+
+    /**
+     * @param annotationValue source of filter properties
+     */
+    public FilterProperties(AnnotationValue<Filter> annotationValue) {
+        this.patterns = annotationValue.stringValues("patterns");
+        this.methods = annotationValue.get("methods", io.micronaut.http.HttpMethod[].class, null);
```

Comment:
Please use `enumValues` it is more efficient

Reply:
@graemerocher I didn't find `enumValues`, there are only functions that return single value